File: src/main/kotlin/com/smody/book/book/api/kakao/KakaoBookApi.java
package com.smody.book.book.api.kakao;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.smody.book.book.api.ApiClient;
import com.smody.book.book.api.BookApi;
import com.smody.book.book.api.BookApiResponse;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Primary;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;

@Primary
@Component
public class KakaoBookApi implements BookApi {

    private final String uri;
    private final String authorization;

    public KakaoBookApi(
            @Value("${api.kakao.uri}") String uri,
            @Value("${api.kakao.authorization}") String authorization) {
        this.uri = uri;
        this.authorization = authorization;
    }

    @Override
    public List<BookApiResponse> findAllByTitle(String title) {
        ApiClient client = ApiClient.builder()
                .uri(uri)
                .addParam("query", title)
                .addHeader("Authorization", authorization)
                .build();

        ResponseEntity<String> response = client.get();
        return parseToBookResponse(response.getBody());
    }

    private List<BookApiResponse> parseToBookResponse(String responseBody) {
        List<KakaoBookApiResponse> kakaoBookApiResponses = new ArrayList<>();
        try {
            kakaoBookApiResponses = new ObjectMapper().readValue(responseBody, KakaoResponse.class).getDocuments();
        } catch (JsonProcessingException e) {
            e.printStackTrace();
        }
        return kakaoBookApiResponses.stream()
                .map(kakaoBookApiResponse -> (BookApiResponse) kakaoBookApiResponse)
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/com/smody/book/book/api/kakao/KakaoBookApi.kt
package com.smody.book.book.api.kakao

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import com.smody.book.book.api.ApiClient
import com.smody.book.book.api.BookApi
import com.smody.book.book.api.BookApiResponse
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.util.*

@Component("kakao")
class KakaoBookApi(
    private val objectMapper: ObjectMapper,
    @param:Value("\${api.kakao.uri}") private val uri: String,
    @param:Value("\${api.kakao.authorization}") private val authorization: String
) : BookApi {
    override fun findAllByTitle(title: String): List<BookApiResponse> {
        val client = ApiClient.builder()
            .uri(uri)
            .addParam("query", title)
            .addHeader("Authorization", authorization)
            .build()
        val response = client.get()
        return parseToBookResponse(response.body)
    }

    private fun parseToBookResponse(responseBody: String?): List<BookApiResponse> {
        var kakaoBookApiResponses: List<KakaoBookApiResponse> = ArrayList()
        try {
            kakaoBookApiResponses =
                objectMapper.readValue(responseBody, KakaoResponse::class.java).documents
        } catch (e: JsonProcessingException) {
            e.printStackTrace()
        }
        return kakaoBookApiResponses
    }
}
